include/onnxruntime/core/framework/provider_options_utils.h

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/common/string_utils.h"
+#include "core/framework/execution_provider.h"
+
+namespace onnxruntime {
+/**
+ * Reads the named provider option.
+ * Returns true if the option is present, false otherwise.
+ */
+template <typename T>
+bool ReadProviderOption(const ProviderOptions& options, const std::string& key, T& value) {
```

Comment:
Is using optional better instead of returning true/false?

Reply:
good idea

---
In reply to: [535835891](https://github.com/microsoft/onnxruntime/pull/5923#discussion_r535835891) [](ancestors = 535835891)

Reply:
actually this signature is useful for the existing usage:
```
CUDAExecutionProviderInfo info{};
ReadProviderOption(options, provider_option_names::kMemLimit, info.cuda_mem_limit);
```
but i can add another function that returns an optional<T>

---
In reply to: [536446380](https://github.com/microsoft/onnxruntime/pull/5923#discussion_r536446380) [](ancestors = 536446380,535835891)

Reply:
Sure. One other thing I noticed is that we validate values returned by this function for only some of the keys. GetOrDefault might be better I guess instead of returning optional as the user will not have the burden to check for the value in optional. Also, it should satisfy all usages.

Reply:
Ah ok, i can change it to ReadProviderOptionOrDefault()

---
In reply to: [536454011](https://github.com/microsoft/onnxruntime/pull/5923#discussion_r536454011) [](ancestors = 536454011)